examples/hdf5.py

```diff
@@ -4,99 +4,199 @@
-        self.limit = 10000 # maximum number of samples to be plotted
+        self.sampler = None
         pg.PlotCurveItem.__init__(self, *args, **kwds)
         
-    def setHDF5(self, data):
-        self.hdf5 = data
-        self.updateHDF5Plot()
-        
+    def setDataSampler(self, sampler):
+        self.sampler = sampler
+        self.loadData()
+
     def viewRangeChanged(self):
-        self.updateHDF5Plot()
-        
-    def updateHDF5Plot(self):
-        if self.hdf5 is None:
+        self.loadData(lazy=True)
+
+    def loadData(self, lazy=False):
+        if self.sampler is None:
             self.setData([])
             return
-        
+
         vb = self.getViewBox()
         if vb is None:
             return  # no ViewBox yet
-        
-        # Determine what data range must be read from HDF5
-        xrange = vb.viewRange()[0]
-        start = max(0,int(xrange[0])-1)
-        stop = min(len(self.hdf5), int(xrange[1]+2))
-        
-        # Decide by how much we should downsample 
-        ds = int((stop-start) / self.limit) + 1
-        
-        if ds == 1:
-            # Small enough to display with no intervention.
-            visible = self.hdf5[start:stop]
-            scale = 1
-        else:
-            # Here convert data into a down-sampled array suitable for visualizing.
-            # Must do this piecewise to limit memory usage.        
-            samples = 1 + ((stop-start) // ds)
-            visible = np.zeros(samples*2, dtype=self.hdf5.dtype)
-            sourcePtr = start
-            targetPtr = 0
-            
-            # read data in chunks of ~1M samples
-            chunkSize = (1000000//ds) * ds
-            while sourcePtr < stop-1: 
-                chunk = self.hdf5[sourcePtr:min(stop,sourcePtr+chunkSize)]
-                sourcePtr += len(chunk)
-                
-                # reshape chunk to be integral multiple of ds
-                chunk = chunk[:(len(chunk)//ds) * ds].reshape(len(chunk)//ds, ds)
-                
-                # compute max and min
-                chunkMax = chunk.max(axis=1)
-                chunkMin = chunk.min(axis=1)
-                
-                # interleave min and max into plot data to preserve envelope shape
-                visible[targetPtr:targetPtr+chunk.shape[0]*2:2] = chunkMin
-                visible[1+targetPtr:1+targetPtr+chunk.shape[0]*2:2] = chunkMax
-                targetPtr += chunk.shape[0]*2
-            
-            visible = visible[:targetPtr]
-            scale = ds * 0.5
-            
-        self.setData(visible) # update the plot
-        self.setPos(start, 0) # shift to match starting index
-        self.resetTransform()
-        self.scale(scale, 1)  # scale to match downsampling
 
-        
+        x1, x2 = vb.viewRange()[0]
+        if not lazy or not self.sampler.isRangeValid(x1, x2):
+            x, y = self.sampler[x1:x2]
+            self.setData(x, y)  # update the plot
+
+
+class DataSampler:
```

Comment:
Ok so I think my main critique here is why does this need to be a class?
All of the "state" being passed at init here is basically used once as an input that you could provide to a lone function and/or bind to a closure that can be called multiple times once you've bound in your base settings.

I would prefer to see all of this done in a set of functions.
If you have a class that defines basically one function that you call (in this case it appears to be `.__getitem__()`), that's not a class.

This class seems to imply there's some kind of "delayed computation" that needs to take place (in a thread that you start right away??)

I don't get why there's all this setting and getting of variables on this object when you can just basically pass it all as input.  If you require delayed computation, use a closure.

Reply:
It's also storing the cached data that gets accessed in subsequent calls to `__getitem__`. How would you suggest storing this instead?

Reply:
well if you want to cache calls to `__getitem__` why not just `lru_cache` it?

Which var is storing the cached data?
It seems the only thing actually being cached is outputs from calls to `._loadlevel()` but those aren't vars here they're implicitly stored inside `lru_cache`.

Reply:
> @lidstrom83 thanks for the shout out 
> 
> Hopefully I'm not overstepping here but personally I would do this a bit differently and make it a little more, "functional" style.

Not at all! I much prefer that as well, if possible.

Reply:
> well if you want to cache calls to `__getitem__` why not just `lru_cache` it?

This would result in a cache entry for potentially every pair of x bounds. Caching by level limits this significantly.

Reply:
So then cache all levels and keep a reference to the func in the parent thread.
Call the now lru cached func ref in the parent thread as needed.